source/core/TypeAndDataManager.cpp

```diff
@@ -739,8 +739,11 @@ Boolean TypeCommon::CompareType(TypeRef otherType)
     if (this == otherType) {
         return true;
     } else if (thisEncoding == kEncoding_Array && otherEncoding == kEncoding_Array) {
-        if (this->Rank() == otherType->Rank())
+        if (this->Rank() != otherType->Rank()) {
```

Comment:
Wow, good catch.  This was a pretty bad bug.

Reply:
For better history tracking, I am moving this change to a separate PR. I will rebase this PR on top of master once the other one is submitted.